File: warpnine_fonts/core/naming.py
"""
Name table manipulation utilities.
"""

from dataclasses import dataclass

from fontTools.ttLib import TTFont

# Copyright text template
COPYRIGHT_TEMPLATE = (
    "Copyright 2020 The Recursive Project Authors (https://github.com/arrowtype/recursive). "
    "Copyright 2014-2021 Adobe (http://www.adobe.com/), with Reserved Font Name 'Source'. "
    "{additional}"
)


@dataclass
class FontNaming:
    """Font naming configuration."""

    family: str  # e.g., "Warpnine Mono"
    style: str  # e.g., "Regular", "Bold"
    postscript_family: str | None = None  # e.g., "WarpnineMono"

    @property
    def full_name(self) -> str:
        """Full font name with family and style."""
        return f"{self.family} {self.style}"

    @property
    def postscript_name(self) -> str:
        """PostScript name (no spaces)."""
        base = self.postscript_family or self.family.replace(" ", "")
        return f"{base}-{self.style.replace(' ', '')}"

    @property
    def unique_id(self) -> str:
        """Unique font identifier."""
        return f"1.0;WARPNINE;{self.postscript_name.replace('-', '')}"


def update_name_table(font: TTFont, naming: FontNaming) -> None:
    """
    Update font name table with consistent naming.

    Args:
        font: TTFont instance to modify
        naming: Naming configuration
    """
    name_table = font["name"]

    for record in name_table.names:
        # nameID 1: Font Family name
        if record.nameID == 1:
            record.string = f"{naming.family} {naming.style}"

        # nameID 3: Unique identifier
        elif record.nameID == 3:
            record.string = naming.unique_id

        # nameID 4: Full font name
        elif record.nameID == 4:
            record.string = naming.full_name

        # nameID 6: PostScript name
        elif record.nameID == 6:
            record.string = naming.postscript_name

        # nameID 16: Typographic Family
        elif record.nameID == 16:
            record.string = naming.family

        # nameID 17: Typographic Subfamily
        elif record.nameID == 17:
            record.string = naming.style


def set_copyright(font: TTFont, additional: str = "") -> None:
    """
    Set copyright notice in name table.

    Args:
        font: TTFont instance to modify
        additional: Additional copyright text to append
    """
    copyright_text = COPYRIGHT_TEMPLATE.format(additional=additional)

    for record in font["name"].names:
        if record.nameID == 0:
            record.string = copyright_text


def update_vf_names(font: TTFont, family: str, postscript_name: str) -> None:
    """
    Update name table for Variable Font.

    Variable fonts don't include style info in some name fields.

    Args:
        font: TTFont instance to modify
        family: Family name (e.g., "Warpnine Mono")
        postscript_name: PostScript name without style (e.g., "WarpnineMono")
    """
    name_table = font["name"]

    for record in name_table.names:
        # nameID 0: Copyright
        if record.nameID == 0:
            copyright_text = COPYRIGHT_TEMPLATE.format(
                additional="Warpnine Mono is based on Recursive Mono Duotone and Noto Sans Mono CJK JP."
            )
            record.string = copyright_text

        # nameID 1: Font Family
        elif record.nameID == 1:
            record.string = family

        # nameID 3: Unique ID
        elif record.nameID == 3:
            record.string = f"1.0;WARPNINE;{postscript_name}"

        # nameID 4: Full Name
        elif record.nameID == 4:
            record.string = family

        # nameID 6: PostScript Name
        elif record.nameID == 6:
            record.string = postscript_name

        # nameID 16: Typographic Family
        elif record.nameID == 16:
            record.string = family

        # nameID 17: Typographic Subfamily
        elif record.nameID == 17:
            record.string = "Regular"

File: crates/cli/src/naming.rs
use std::path::Path;

use anyhow::Result;

use crate::font_ops::{map_name_records, modify_font_in_place};

const COPYRIGHT_TEMPLATE: &str = "Copyright 2020 The Recursive Project Authors (https://github.com/arrowtype/recursive). \
Copyright 2014-2021 Adobe (http://www.adobe.com/), with Reserved Font Name 'Source'. ";

#[derive(Debug, Clone)]
pub struct FontNaming {
    pub family: String,
    pub style: String,
    pub postscript_family: Option<String>,
    pub copyright_extra: Option<String>,
}

impl FontNaming {
    pub fn full_name(&self) -> String {
        format!("{} {}", self.family, self.style)
    }

    pub fn postscript_name(&self) -> String {
        let base = self
            .postscript_family
            .clone()
            .unwrap_or_else(|| self.family.replace(' ', ""));
        format!("{base}-{}", self.style.replace(' ', ""))
    }

    pub fn unique_id(&self) -> String {
        let ps_name = self.postscript_name().replace('-', "");
        format!("1.0;WARPNINE;{ps_name}")
    }

    pub fn copyright(&self) -> String {
        match &self.copyright_extra {
            Some(extra) => format!("{COPYRIGHT_TEMPLATE}{extra}"),
            None => COPYRIGHT_TEMPLATE.to_string(),
        }
    }
}

pub fn set_name(path: &Path, naming: &FontNaming) -> Result<()> {
    modify_font_in_place(path, |font, builder| {
        let name = map_name_records(font, |name_id, _current| match name_id {
            0 => Some(naming.copyright()),
            1 => Some(format!("{} {}", naming.family, naming.style)),
            3 => Some(naming.unique_id()),
            4 => Some(naming.full_name()),
            6 => Some(naming.postscript_name()),
            16 => Some(naming.family.clone()),
            17 => Some(naming.style.clone()),
            _ => None,
        })?;
        builder.add_table(&name)?;
        Ok(())
    })?;

    println!(
        "{}: set name to '{}' ({})",
        path.file_name().unwrap_or_default().to_string_lossy(),
        naming.full_name(),
        naming.postscript_name()
    );

    Ok(())
}
